src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -84,12 +84,82 @@ namespace smt::noodler {
+                        if (!right_side_vars.empty() || element_to_process.length_sensitive_vars.count(var) > 0) {
+                            assert(substitution_map.count(var) == 0 && element_to_process.aut_ass.count(var) > 0);
+                            // we prepare substitution for all vars on the left or only the length vars on the right (as non-length vars are probably not needed? TODO: would it make sense to update non-length vars too?)
+                            substitution_map[var] = {};
+                            element_to_process.aut_ass.erase(var);
+                        }
+                    } else {
+                        non_empty_side_contains_empty_word = false;
+                        break;
+                    }
+                }
+                if (!non_empty_side_contains_empty_word) {
+                    // in the case that the non_empty side does not contain empty word
+                    // the inclusion cannot hold (noodlification would not create anything)
+                    continue;
+                }
+
+                // TODO: all this following shit is done also during normal noodlification, I need to split it to some better defined functions
+
+                // we might remove some variables from the left side
+                // so we need to add all nodes whose variable assignments are going to change on the right side (i.e. we follow inclusion graph) for processing
+                for (const auto &node : element_to_process.inclusion_graph->get_edges_from(node_to_process)) {
+                    // we push only those nodes which are not already in nodes_to_process
+                    // if the node_to_process is on cycle, we need to do BFS
+                    // if it is not on cycle, we can do DFS
+                    // TODO: can we really do DFS??
+                    element_to_process.push_unique(node, is_node_to_process_on_cycle);
+                }
+
+                // we need to make a deep copy, because we will be updating this graph
+                auto new_node_to_process = element_to_process.make_deep_copy_of_inclusion_graph_only_nodes(node_to_process);
+                //remove processed inclusion from the inclusion graph
+                element_to_process.inclusion_graph->remove_node(new_node_to_process); // TODO: is this safe?
+                // do substitution in the inclusion graph
+                element_to_process.substitute_vars(substitution_map);
+                // update inclusion graph edges
+                element_to_process.inclusion_graph->add_inclusion_graph_edges();
+                // update the substitution_map of new_element by the new substitutions
+                element_to_process.substitution_map.merge(substitution_map);
+
+                // TODO: should we really push to front when not on cycle?
+                // TODO: maybe for this case of one side being empty, we should just push to front? 
```

Comment:
Could we try to resolve these TODOs now? They are both valid questions, in my opinion.

Reply:
Right now I am doing what Vojta was doing in Noodler, if it is chain-free (or more specifically, we are in the "chain-free part" if the inclusion graph), then I do DFS, otherwise I do BFS. I think DFS is better to get to some solution faster, but we should maybe leave it for the meeting.